src/IO/Observer/Actions/RegisterEvents.hpp

```diff
@@ -0,0 +1,98 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "IO/Observer/Actions/ObserverRegistration.hpp"
+#include "IO/Observer/ArrayComponentId.hpp"
+#include "IO/Observer/ObserverComponent.hpp"
+#include "IO/Observer/Tags.hpp"
+#include "IO/Observer/TypeOfObservation.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "ParallelAlgorithms/EventsAndTriggers/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
+
+namespace observers {
+namespace Actions {
+namespace detail {
+CREATE_HAS_TYPE_ALIAS(observation_registration_tags)
+CREATE_HAS_TYPE_ALIAS_V(observation_registration_tags)
+}  // namespace detail
+
+/*!
+ * \brief Registers itself with the local observer parallel component so the
+ * observer knows to expect data from this component, and also whether to expect
+ * Reduction, Volume, or both Reduction and Volume data observation calls.
+ */
+struct RegisterEventsWithObservers {
+  template <typename DbTagList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagList>&&> apply(
+      db::DataBox<DbTagList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    auto& observer =
+        *Parallel::get_parallel_component<observers::Observer<Metavariables>>(
+             cache)
+             .ckLocalBranch();
+    std::vector<
+        std::pair<observers::TypeOfObservation, observers::ObservationId>>
+        type_of_observation_and_observation_id_pairs;
+
+    const auto& triggers_and_events =
+        db::get<::Tags::EventsAndTriggersBase>(box);
+    using possible_events = typename std::decay_t<decltype(
+        triggers_and_events)>::event_type::creatable_classes;
+    for (const auto& trigger_and_events :
+         triggers_and_events.events_and_triggers()) {
+      for (const auto& event : trigger_and_events.second) {
+        tmpl::for_each<
+            possible_events>([&box, &event,
+                              &type_of_observation_and_observation_id_pairs](
+                                 auto event_type_v) noexcept {
+          using EventType = typename decltype(event_type_v)::type;
+          if constexpr (detail::has_observation_registration_tags_v<
+                            EventType>) {
+            const auto* const derived_class_ptr =
```

Comment:
Clearly I need to improve the documentation. Here's what's happening:

Events can do anything, not just observations, so we need a way of getting the list of/looping over events that will be doing observations. We do this first getting the list of all events that are allowed. Then we loop over all triggers, for each trigger, loop over all the events. We then loop over the possible events and, if that event has a member function that specifies the observation ID needed for registration, try to downcast the current base class event to that derived class. If that works, we perform the registration of the event, if not, then we don't.

Some things that I think are in general necessary to support in registration:
- any number of events
- the observation IDs may depend on something in the DataBox
- this has nothing to do with how complicated the inheritance change is and everything to do with whether the final class in the chain (or the one being used in the factory to create events) has the member function `get_observation_type_and_id_for_registration`. So nontrivial inheritance is fine
- I don't understand what you are thinking with a free function. I don't see how that would work since the observation ID depends on both DataBox info and runtime arguments passed in from the input file.

Reply:
Ah, okay, that explanation helps a great deal.

Let me see if I can make more concrete my inheritance concern, though perhaps it is a case that should not come up in practice -- at any rate, if our design prohibits it, we should document the constraint.
Consider the scenario in which we have two events `BaseEvent` and `DerivedEvent`, such that `BaseEvent` defines a `get_observation_type_and_id_for_registration`, and `DerivedEvent` inherits the function, and both `BaseEvent` and `DerivedEvent` are in the `possible_events`, but the input file specifies only a single `DerivedEvent` member to be triggered. Now my understanding of this code is (and I'm happy to be corrected):
- the downcast will succeed both when `EventType` is `DerivedEvent` and when it is `BaseEvent` (and both will appear in the loop over `possible_events`
- the registration will occur twice, calling the `BaseEvent` version of the member `get_observation_type_and_id_for_registration` once, and the `DerivedEvent` version once (unless it's a virtual function...).
Am I mistaken?

Reply:
Sure, I view this as unsupported behavior. It would be reasonably easy to check that only one downcast is possible per class being registered.

Reply:
That sounds good -- if you could add an error in the case that multiple downcasts succeed, I think that would satisfy my concerns.